packages/beacon-node/src/chain/archiver/pruneHistory.ts

```diff
@@ -0,0 +1,73 @@
+import {ChainConfig} from "@lodestar/config";
+import {computeStartSlotAtEpoch} from "@lodestar/state-transition";
+import {Epoch} from "@lodestar/types";
+import {Logger} from "@lodestar/utils";
+import {IBeaconDb} from "../../db/interface.js";
+import {Metrics} from "../../metrics/index.js";
+
+/**
+ * The maximum number of blocks to delete in a single batch
+ *
+ * Too much and the main thread will block
+ */
+// TODO tune these with metrics/data
+const MAX_BLOCKS_TO_DELETE = 4096;
+
+/**
+ * The maximum number of states to delete in a single batch
+ *
+ * Too much and the main thread will block
+ */
+const MAX_STATES_TO_DELETE = 256;
+
+export async function pruneHistory(
+  config: ChainConfig,
+  db: IBeaconDb,
+  logger: Logger,
+  metrics: Metrics | null | undefined,
+  finalizedEpoch: Epoch,
+  currentEpoch: Epoch,
+  maxBlocksToDelete = MAX_BLOCKS_TO_DELETE,
+  maxStatesToDelete = MAX_STATES_TO_DELETE
+): Promise<void> {
+  const blockCutoffEpoch = Math.min(
+    // set by config, with underflow protection
+    Math.max(currentEpoch - config.MIN_EPOCHS_FOR_BLOCK_REQUESTS, 0),
+    // ensure that during (extremely lol) long periods of non-finality we don't delete unfinalized epoch data
+    finalizedEpoch
+  );
+  const blockCutoffSlot = computeStartSlotAtEpoch(blockCutoffEpoch);
+
+  logger.debug("Preparing to prune history", {
+    currentEpoch,
+    finalizedEpoch,
+    blockCutoffEpoch,
+  });
+
+  const step0 = metrics?.pruneHistory.fetchKeys.startTimer();
+  const [blocks, states] = await Promise.all([
+    db.blockArchive.keys({gte: 0, lt: blockCutoffSlot, limit: maxBlocksToDelete}),
+    db.stateArchive.keys({gte: 0, lt: finalizedEpoch, limit: maxStatesToDelete}),
+  ]);
```

Comment:
Seems good in general. but once lol scenario is resolved there could be a very long list of keys to be deleted. What if it's `> MAX_BLOCKS_TO_DELETE`. 

Should we not rather choose to `MAX_BLOCKS_TO_KEEP` than the otherway around.

Reply:
I thought about just removing this whole limit thing as a premature optimization